File: packages/python/mythos_sdk/jwks_cache.py
import time
from typing import Any
from urllib.parse import quote

from .http import get_http_client

CACHE_TTL = 600  # 10 minutes

_cache_by_url: dict[str, dict[str, Any]] = {}
_fetched_at_by_url: dict[str, float] = {}


def _is_stale(api_url: str) -> bool:
    fetched_at = _fetched_at_by_url.get(api_url)
    if fetched_at is None:
        return True
    return time.monotonic() - fetched_at > CACHE_TTL


async def _fetch_jwks(api_url: str) -> dict[str, Any]:
    client = get_http_client()
    resp = await client.get(f"{api_url}/.well-known/jwks.json")
    resp.raise_for_status()
    data: dict[str, Any] = resp.json()
    _cache_by_url[api_url] = data
    _fetched_at_by_url[api_url] = time.monotonic()
    return data


async def get_jwks(api_url: str, force_refresh: bool = False) -> dict[str, Any]:
    cached = _cache_by_url.get(api_url)
    if not force_refresh and cached is not None and not _is_stale(api_url):
        return cached
    return await _fetch_jwks(api_url)


async def get_jwks_with_kid_fallback(api_url: str) -> dict[str, Any]:
    return await get_jwks(api_url, force_refresh=True)


def clear_cache() -> None:
    _cache_by_url.clear()
    _fetched_at_by_url.clear()

File: packages/node/src/jwks-cache.ts
import { createLocalJWKSet, type JSONWebKeySet } from 'jose';
import { mythosRequest } from './http';

const CACHE_TTL_MS = 10 * 60 * 1000;

interface CacheEntry {
  keySet: ReturnType<typeof createLocalJWKSet>;
  fetchedAt: number;
}

const _cache = new Map<string, CacheEntry>();

function isStale(entry: CacheEntry): boolean {
  return Date.now() - entry.fetchedAt > CACHE_TTL_MS;
}

async function fetchJwks(apiUrl: string): Promise<ReturnType<typeof createLocalJWKSet>> {
  const res = await mythosRequest(`${apiUrl}/.well-known/jwks.json`, { method: 'GET' });
  if (!res.ok) {
    throw new Error(`JWKS fetch failed: ${res.status}`);
  }
  const jwks = (await res.json()) as JSONWebKeySet;
  const keySet = createLocalJWKSet(jwks);
  _cache.set(apiUrl, { keySet, fetchedAt: Date.now() });
  return keySet;
}

export async function getKeySet(apiUrl: string, forceRefresh = false): Promise<ReturnType<typeof createLocalJWKSet>> {
  const cached = _cache.get(apiUrl);
  if (!forceRefresh && cached && !isStale(cached)) {
    return cached.keySet;
  }
  return fetchJwks(apiUrl);
}

export async function getKeySetWithKidFallback(apiUrl: string): Promise<ReturnType<typeof createLocalJWKSet>> {
  return getKeySet(apiUrl, true);
}

export function clearCache(): void {
  _cache.clear();
}
